### ISSUE TITLE
oom

### ISSUE THREAD

**Issue by Pydataman** [shivammehta007/Neural-HMM#1]
gpu memory:24G

File "src/model/HMMComponents/EmissionModel.py", line 45, in forward
    out = emission_dists.log_prob(x_t)
  File "/lib/python3.7/site-packages/torch/distributions/normal.py", line 77, in log_prob
    return -((value - self.loc) ** 2) / (2 * var) - log_scale - math.log(math.sqrt(2 * math.pi))

**Comment by shivammehta007** [comment]
Hello! 
OOM is not a code issue. It is a known problem because of the way PyTorch handles computation graphs the forward algorithm makes a huge gradient tape. We are working on ideas to mitigate it. Please feel free to suggest something. 

Meanwhile, things you can do to mitigate it is to do mix-precision training by changing the `precision=16` in `src/hparams.py`. Also the length of utterance is a major factor as this is an autoregressive architecture. You can try clipping the utterances or reducing the batch size. 
You can also try a monostate model by changing the `state_per_phone=1` in `src/hparams.py`, which generates decent comparable results in preliminary experiments.

**Comment by shivammehta007** [comment]
Also, If you are training on a single GPU. You can clone the branch  [gradient_checkpoinint](https://github.com/shivammehta007/Neural-HMM/tree/gradient_checkpointing), it utilises the gradient checkpointing feature of PyTorch and prevents OOM Error.

Hope this solves the question. Feel free to open this issue again if something you feel is not right or add more comments to this thread.
